### PR TITLE
make task-name as an net-alias to the backing container

### LINKED ISSUE #24466: docker 1.12 swarm mode: how to connect to another container on overlay network?
Sorry if this is a duplicate.  I feel like I'm asking a dumb question but I couldn't figure this out even with all the documentation and tutorials.  I think it would be great if the answer went into another page of [this tutorial](https://docs.docker.com/engine/swarm/swarm-tutorial/).

In docker 1.12.0-rc3 I set up 3 nodes in swarm mode and an overlay network.  

I notice the task names are predictably networktest.1, .2, .3:
```
$ docker service tasks networktest
ID                         NAME           SERVICE      IMAGE   LAST STATE         DESIRED STATE  NODE
d6naj58s8izkb822fpd4b6tl0  networktest.1  networktest  alpine  Running 4 seconds  Running        ip-172-31-47-32.us-west-2.compute.internal
7sglm20hhbmn1ws5fiqvqfgx5  networktest.2  networktest  alpine  Running 4 seconds  Running        ip-172-31-43-158.us-west-2.compute.internal
au4unlfyavw5y6vhodaw2dwe4  networktest.3  networktest  alpine  Running 4 seconds  Running        ip-172-31-43-158.us-west-2.compute.internal
```
but the container names have a random suffix:
```
[ec2-user@ip-172-31-47-32 ~]$ docker ps
CONTAINER ID        IMAGE                           COMMAND                  CREATED             STATUS              PORTS                                  NAMES
e67110a596b0        alpine:latest                   "sleep 36000"            5 minutes ago       Up 5 minutes                                               networktest.1.d6naj58s8izkb822fpd4b6tl0
```

and
```
[ec2-user@ip-172-31-43-158 ~]$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
412ed90a1cd7        alpine:latest       "sleep 36000"       4 minutes ago       Up 3 minutes                            networktest.2.7sglm20hhbmn1ws5fiqvqfgx5
44ea273ba889        alpine:latest       "sleep 36000"       4 minutes ago       Up 3 minutes                            networktest.3.au4unlfyavw5y6vhodaw2dwe4
```

How are we then supposed to have container-to-container communication?

So giving that full randomly-generated name allows one container to ping another:
```
[ec2-user@ip-172-31-47-32 ~]$ docker exec -it e67110a596b0 ping networktest.2.7sglm20hhbmn1ws5fiqvqfgx5
PING networktest.2.7sglm20hhbmn1ws5fiqvqfgx5 (10.0.0.4): 56 data bytes
64 bytes from 10.0.0.4: seq=0 ttl=64 time=0.319 ms
64 bytes from 10.0.0.4: seq=1 ttl=64 time=0.236 ms
64 bytes from 10.0.0.4: seq=2 ttl=64 time=0.217 ms
```

*But* I really wish this worked too:
```
[ec2-user@ip-172-31-47-32 ~]$ docker exec -it e67110a596b0 ping networktest.2
ping: bad address 'networktest.2'
```

I didn't see any parameters to `docker service create` for `links`, so how do I configure one container to communicate with another? 

In [this blog](https://blog.docker.com/2016/06/docker-1-12-built-in-orchestration/) it implied the `frontend` and `redis` database would be connected with just the following commands. Was that misleading?
```
docker network create -d overlay mynet
docker service create --name frontend --replicas 5 -p 80:80/tcp --network mynet mywebapp
docker service create --name redis --network mynet redis:latest
```

### PR DESCRIPTION
**Pull request by mavenugo**
fixes https://github.com/docker/docker/issues/24466

This will help with resolving DNS names using task-name

Signed-off-by: Madhu Venugopal madhu@docker.com

### CODE REVIEW

**Comment by mavenugo** [comment]
ping @tiborvass @thaJeztah . Do you think this is a useful thing to have ? I could not decide it myself if it is useful for 1.12 or nice to have....

**Comment by thaJeztah** [comment]
So, for reference; given a service `web`;

```
docker service tasks web

ID                         NAME   SERVICE  IMAGE         LAST STATE                  DESIRED STATE  NODE
9tsk9hw2uac4tsharxkuzf981  web.1  web      nginx:alpine  Running about a minute ago  Running        moby
1opisbjf2ypok3cakoq2gmm4e  web.2  web      nginx:alpine  Running 6 seconds ago       Running        moby
3zqyv3sx6juzmlvxrcpl7adxo  web.3  web      nginx:alpine  Running 6 seconds ago       Running        moby
```

```
docker ps

CONTAINER ID        IMAGE               COMMAND                  CREATED              STATUS              PORTS               NAMES
561e2a934d6f        nginx:alpine        "nginx -g 'daemon off"   About a minute ago   Up 59 seconds       80/tcp, 443/tcp     web.2.1opisbjf2ypok3cakoq2gmm4e
fa73bbcbf766        nginx:alpine        "nginx -g 'daemon off"   About a minute ago   Up 59 seconds       80/tcp, 443/tcp     web.3.3zqyv3sx6juzmlvxrcpl7adxo
bec6bc5a579a        nginx:alpine        "nginx -g 'daemon off"   2 minutes ago        Up 2 minutes        80/tcp, 443/tcp     web.1.9tsk9hw2uac4tsharxkuzf981
``` 

Individual task-containers can currently be discovered by their *full* (container) name (e.g.`web.1.9tsk9hw2uac4tsharxkuzf981`), but not the *task* name that appears in `docker service tasks` (e.g. `web.1`); this change adds the task name `web.1` as an alias.

Without this change, I think the only way to find the container that's related to a task is to `docker inspect <task-id>`, and get the container-id from that output, so I can see this being helpful.

It's related to https://github.com/docker/docker/issues/24241, however, so the same concern could be raised https://github.com/docker/docker/issues/24241#issuecomment-232237105, so:


ping @aluzzardi @dnephin @stevvooe PTAL

**Comment by thaJeztah** [comment]
Added to the milestone to make it easier to track, but we can move it or close it depending on what we think is best

**Comment by thaJeztah** [comment]
Oh, this fixes https://github.com/docker/docker/issues/24466, so added that to the description; and contains a use-case / more background

File: daemon/cluster/executor/container/container.go [hunk 1]
```diff
@@ -81,6 +81,16 @@ func (c *containerConfig) name() string {
...
 }
 
+func (c *containerConfig) shortName() string {
+	if c.task.Annotations.Name != "" {
+		// if set, use the container Annotations.Name field, set in the orchestrator.
+		return c.task.Annotations.Name
+	}
+
+	// fallback to service.slot.id.
+	return strings.Join([]string{c.task.ServiceAnnotations.Name, fmt.Sprint(c.task.Slot)}, ".")
```
Comment (dnephin) [inline daemon/cluster/executor/container/container.go]: Minor nit:

```go
return fmt.Sprintf("%s.%d", c.task.ServiceAnnotations.Name, c.task.Slot)
```
would make this a little more obvious

**Comment by dnephin** [comment]
I don't think this is all that related to #2424.

Providing the task short id as a discoverable alias sounds good to me. In the case where a name is set, could we still add the task name as an alias? Isn't the name already discoverable by default?

**Comment by mavenugo** [comment]
@dnephin @thaJeztah addressed your comments. PTAL.

**Comment by thaJeztah** [comment]
LGTM

/cc-ing @sfsmithcha in case we want to document this somewhere (although I'm not sure where directly)

File: daemon/cluster/executor/container/container.go [hunk 2]
```diff
@@ -77,8 +77,18 @@ func (c *containerConfig) name() string {
 		return c.task.Annotations.Name
 	}
 
-	// fallback to service.slot.id.
-	return strings.Join([]string{c.task.ServiceAnnotations.Name, fmt.Sprint(c.task.Slot), c.task.ID}, ".")
+	// fallback to service.slot.id
+	return fmt.Sprintf("%s.%s", c.shortName(), c.task.ID)
```
Comment (dnephin) [inline daemon/cluster/executor/container/container.go]: Sorry, I think my comment was misleading.  Wont this use a name of `service.slot.taskid` ?

Comment (dnephin) [inline daemon/cluster/executor/container/container.go]: I was thinking that every task would get the following aliases:
* container id
* task name (service.slot)

(as well as the service name which is load balanced)

File: daemon/cluster/executor/container/container.go [hunk 3]
```diff
@@ -77,8 +77,18 @@ func (c *containerConfig) name() string {
...
+}
+
+func (c *containerConfig) shortName() string {
+	if c.task.Annotations.Name != "" {
+		// if set, use the container Annotations.Name field, set in the orchestrator.
+		return c.task.Annotations.Name
+	}
+
+	// fallback to service.slot
+	return fmt.Sprintf("%s.%s", c.task.ServiceAnnotations.Name, fmt.Sprint(c.task.Slot))
```
Comment (stevvooe) [inline daemon/cluster/executor/container/container.go]: Make `name()` use this. Also, document the method.

**Comment by stevvooe** [comment]
@mavenugo Isn't this using a completely different method for generating the DNS name? Should we not have this all in the same place? What are the other projections of task into the dns space?

File: daemon/cluster/executor/container/container.go (hunk 2 above)
Comment (mavenugo) [inline daemon/cluster/executor/container/container.go]: @dnephin you are looking at `name()` which is infact `service.slot.taskid`. whereas the shortName is `service.slot` which is the task-name and it is used as an net-alias.

In short, yes a container will have container-id and task-name as net-aliases. Ofcourse service-name is used for load-balancing.

File: daemon/cluster/executor/container/container.go (hunk 3 above)
Comment (mavenugo) [inline daemon/cluster/executor/container/container.go]: yes. `name()` is infact using this method above.

File: daemon/cluster/executor/container/container.go (hunk 2 above)
Comment (dnephin) [inline daemon/cluster/executor/container/container.go]: Thanks, makes sense

**Comment by stevvooe** [comment]
Also, how are aliases on the network attachment propagated?

**Comment by mavenugo** [comment]
@stevvooe Sorry. I dont quite get your comment. The only addition that this change brings in is the additional dns entry for a container with the task-name. So with this change, a container can be reached by its

* container-name (service.slot.taskid)
* task-name (service.slot)
* container-id
* service-name (via LB)

container-name is the primary DNS entry for the container.
task-name & container-id are aliases.
service-name is the Virtual-name for the container.

**Comment by mavenugo** [comment]
@stevvooe the aliases are propagated the exact same way all other names are propagated (via Gossip SD mechanism).

**Comment by mavenugo** [comment]
If this needs more discussions, we can consider it for later releases. I dont see this as mandatory for 1.12.0 release.

**Comment by cpuguy83** [comment]
Removing from the milestone, setting to design review.

**Comment by toughIQ** [comment]
This feature is useful, if you have to create your own Loadbalancer service (with different balancing strategies than swarm provides) and need to know the backend tasks.

At first I thought, that `service.slot-id` would suffice, but this is only the case, if the number of backend tasks is fixed and gets not changed by `scale` command. Or your LB is set up with eg. `service.1` to `service.100` as backends by default and only uses the currently running serivce-ids.

Example: you start with 3 replicas. So you get `service.1`, `service.2` and `service.3`. That's predictable. If you then scale up to 10, you have service IDs from 1-10. Also some kind of predictable. But if you scale down to 3 again, you are left with 3 tasks, but their slot IDs can be anything from 1-10 and you cannot say, what it will be.

In my current project https://github.com/toughIQ/docker-maxscale I used the dnsRR (by now undocumented feature) results of `gentent hosts tasks.SERVICE` to create a dynamic config at startup.

**Comment by stevvooe** [comment]
@toughIQ For something like that, I'd really like to see something like [`DNAME`](https://en.wikipedia.org/wiki/CNAME_record#DNAME_record).

In general, it is not critical that the load balancer is _always_ synced with the current number of tasks if the load balancer is health checking backends. Even in the case where there replica holes, the load balancer will be able to work out which ones are up.

The other option here would be to have it consume scaling events.

**Comment by stevvooe** [comment]
@mavenugo To be compatible with the hierarchy we've been discussing, these really need to be the other way around:

> container-name (service.slot.taskid)

taskid.slot.service

> task-name (service.slot)

slot.service

> container-id

container-id.service

> service-name (via LB)

service.namespace.cluster

This will allow us to combine DNS spaces in the future in interesting ways.

**Comment by thaJeztah** [comment]
@stevvooe can you open a separate issue for that for discussion? Also; can we make that change after 1.12 without breaking backward compatibility?

**Comment by stevvooe** [comment]
@thaJeztah Filed docker/swarmkit#1242.

**Comment by cpuguy83** [comment]
Not sure how useful this would be without having some way to discover how many slots there are.

**Comment by youngnicks** [comment]
This change would help dockerize apps that cluster themselves together, such as Zookeeper. With the current version of Zookeeper, you need to specify other servers in the cluster for it to connect to. With this change you could have the following in the zoo.cfg:
```
server.1=zoo.1:2888:3888
server.2=zoo.2:2888:3888
server.3=zoo.3:2888:3888
```

**Comment by stevvooe** [comment]
@mavenugo @youngnicks We are looking at a comprehensive solution here. These use cases are a particular pain point and we need to address them.

**Comment by thaJeztah** [comment]
@mavenugo @stevvooe should we close this while you're working on an alternative?

**Comment by stevvooe** [comment]
@thaJeztah Until there is an actual plan here, we should not be changing the service mapping into DNS.

**Comment by stevvooe** [comment]
@mavenugo This needs a proposal around docker/swarmkit#1242. We can't keep adding bespoke DNS entries without documenting their guarantees. For example, I've see slots proposed as a method of identifying instances consistently, but we need to make sure that slots have the right guarantees for that to work and we comfortable supporting that.

I'm not wholly opposed to this addition, so I don't think we need to close this, but we do need to have a real plan for DNS-based service discovery.

**Comment by mavenugo** [comment]
apologies on the delay. Yes, I agree that we should have a complete proposal on this. lets continue the discussion in https://github.com/docker/swarmkit/issues/1242.

**Comment by LK4D4** [comment]
@mavenugo can we close this? We have too many PRs now :)

**Comment by mavenugo** [comment]
@LK4D4 yes... closing this.

**Comment by eyz** [comment]
This sounds very useful, as mentioned, for mapping in external service load balancers.

**Pull request closed without merging**